File: app/src/main/java/com/absinthe/anywhere_/adapter/gift/ChatAdapter.java
package com.absinthe.anywhere_.adapter.gift;

import com.chad.library.adapter.base.BaseNodeAdapter;
import com.chad.library.adapter.base.entity.node.BaseNode;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class ChatAdapter extends BaseNodeAdapter {

    public static final int TYPE_LEFT = 0;
    public static final int TYPE_RIGHT = 1;
    public static final int TYPE_INFO = 2;

    public ChatAdapter() {
        addNodeProvider(new LeftChatProvider());
        addNodeProvider(new RightChatProvider());
        addNodeProvider(new InfoProvider());
    }

    @Override
    protected int getItemType(@NotNull List<? extends BaseNode> list, int i) {
        BaseNode node = list.get(i);

        if (node instanceof LeftChatNode) {
            return TYPE_LEFT;
        } else if (node instanceof RightChatNode) {
            return TYPE_RIGHT;
        } else if (node instanceof InfoNode) {
            return TYPE_INFO;
        }
        return -1;
    }
}

File: app/src/main/java/com/absinthe/anywhere_/adapter/gift/ChatAdapter.kt
package com.absinthe.anywhere_.adapter.gift

import com.chad.library.adapter.base.BaseNodeAdapter
import com.chad.library.adapter.base.entity.node.BaseNode

class ChatAdapter : BaseNodeAdapter() {

    init {
        addNodeProvider(LeftChatProvider())
        addNodeProvider(RightChatProvider())
        addNodeProvider(InfoProvider())
    }

    override fun getItemType(data: List<BaseNode>, position: Int): Int {
        return when (data[position]) {
            is LeftChatNode -> {
                TYPE_LEFT
            }
            is RightChatNode -> {
                TYPE_RIGHT
            }
            is InfoNode -> {
                TYPE_INFO
            }
            else -> -1
        }
    }

    companion object {
        const val TYPE_LEFT = 0
        const val TYPE_RIGHT = 1
        const val TYPE_INFO = 2
    }
}